### ISSUE TITLE
exports name same as instance of Class 

### ISSUE THREAD

**Issue by DannyLingard** [deltaepsilon/quiver-functions#1]
Hi Chris,

Is there a particular reason you named exports in index.js the same as the stored class instance, i.e `exports.updateUser`? Or is this purely decorative for consistency.

I've also noticed that the constructor runs against each export, even if that export doesn't reference the instantiated class. Have you noticed this and is it expected?

**Comment by deltaepsilon** [comment]
Yeah. That naming is all for consistency. The only place that it matters is for https functions, because you do end up referencing them by name. So ```exports.environment``` will map to ```/environment```.

I don't understand what you mean by the constructor running against each export. Could you refer to me to the line(s) of code in question? I've been using these functions on my current project, and they've been behaving as expected... but I'd love to catch bugs early!

**Comment by DannyLingard** [comment]
Cool thanks for that. I haven't used your functions yet but have adopted a lot of the principles (as well as from YouTube vids). I'll will be using the login and user create once I get to that point. I'm a wee way off the testing and environments part.

If you console.log in the constructor of any `new` classes in index.js, that log will appear against all function.exports in the logs, even  if that class isn't referenced by the function.

I stupidly also didn't realise that the classes in  `index.js` are instantiated only when functions are deployed, not at the write event. My classes got so large and intertwined that the promise chains became a nightmare to manage and debug.

Rather than rely on promise chains, I'm instead writing to a `/jobs/` ref in an attempt to create single purpose classes. I then have a class handler for each entity type which I instantiate in `index.js` and reference CRUD methods:
`functions.database.ref('/jobs/create-entity/{userId}/{entityId}').onWrite(entityHandler.create())`

The above is identical to your set up in that the database config etc is passed in. Only difference is that in the create() method, I instantiate an instance of a specific class which purely undertakes that create, allowing full use of this and constructor to share across methods within that class. Once complete, I then write to a different `/jobs/` ref if additional work is needed, eg I need to reference the relationship keys. Good news is I can keep passing through an ID from an initial client write to the `/queue/` ref which essentially allows to kick off a batch of related async tasks.

This is what's worked for me and what I found manageable in terms of OOP relating to entities. Not sure if it's the right way or the best way but I found if/else promise chaining hard to follow even when split into different methods. 

Also of note, the reason my classes got so big is because I have global entity objects (in my FB DB) and user entities (read only to the client). I'm storing all properties to the global and select properties to the user entity. I also have a full entity object, a lookup entity object, i.e entity -> push().key: true and query entity object to indexOn e.g entity -> push().key: "some string that can't be a key". 

To manage all this, I use Object.defineProperties() to define getters and setters for validation of internal members of entity specific objects. Down side is this is a lot of atomic updates at each property which creates bloat but better visibility that returning an object.

Cheers for the work you put in as well, your the only dependable source I've found other than Firebase themselves.

**Comment by deltaepsilon** [comment]
It sounds like you're on a great path. I'm developing these functions for my current project, and I figured I might as well break out any generic work that I do into a separate project so that it can a) be open source and b) I can easily reuse them later.

We're still so early in adoption of Functions that I don't think anyone knows if they'll be reusable across projects, or if it'll be more efficient to copy/paste and use them for inspiration. 

I like your ```/jobs/``` architecture! That's a super robust way to set up your workflows, and keeping functions single-purpose is always nice. It's a balance between the extra complexity of having more, smaller functions, or fewer large functions. I'm still learning where on that continuum I want to be. I'm leaning toward smaller functions. But I'm not ready to commit to making each one entirely single-purpose... at least not yet.

We're on the bleeding edge of this technology, so we should continue to compare notes. Do you Slack? Firebase has a [community channel](https://firebase.googleblog.com/2016/02/come-join-firebase-slack-community_13.html) where we could start a chat thread.

**Comment by DannyLingard** [comment]
Yeah I definitely agree it's on a project by project basis in terms of which path suits. Mines getting large in terms of entity interactions, timing of async flows from certain entities (both internal queries and external API) and data fan out. It does come with a lot of boilerplate though and the separation means not everything is immediately clear without sufficient digging.

I'm almost contemplating a `/views` ref to store all of my public/per user entity objects. I was originally storing each entity under `/user-entity/{uid}/` but if I align that fully with user actions, then I lose the user version of those entities in the global format which I may want later. An example being, if a user delete's an entity, at the global `/entity/` ref I have a `userDeleted` property which I set to true, then duplicate to `/deleted-entity` and `/deleted-entities/{id} = true`. 

However, if I apply this to my user refs, then I'll be pulling unnecessary data down. So in that sense, I only want to send the 'live' data to the client and only the required fields for view layer as you can't exclude fields when receiving snaps. All I can think it's to duplicate again at yet another ref so that my database looks like:

 - /jobs/ (Single purpose per entity crud operation, joining entities and dispatch dependant cross entity jobs)
 - /queue/ (User writes are checked, validated and dispatched to relevant jobs ref
 - /view/ (global and per user view of what only needs to be sent down stream)
 - /user/ (Storing user and all user based entity keys)
 - /entity/ (Some object dictionary)
 - /entities/ (Lookup table of above entity IDs)
 - /entities-lookup/ (Alternate lookup table for queiries where key isn't known and value cannot be stored as a key)
..........
the entity structure is replicated with /user-entity/...

I don't currently Slack but pretty keen to jump on board the community channel soon to compare notes etc.

**Comment by deltaepsilon** [comment]
Oh goodness, your data structure is getting tricky.

I'm not sure there's anything to do about that though... sometimes one gets to just suck it up, make a ton or writes, write a ton of tests and embrace the complexity.

You obviously know what you're doing, but I'll share my preferred data structure anyway :)

- admin/{objectType}/{objectKey}
- /users/{uid}
- /userOwned/{objectType}/{uid}
- /userWriteable/{objectType}/{uid}
- /userReadable/{objectType}/{uid}

This minimizes the rules that I have to write. But I end up with the same number of endpoints... I just hate writing and maintaining rules.

**Comment by DannyLingard** [comment]
I really like that data structure actually. I was going as flat as possible but the above get's a lot of wins with the cascading nature of rules, allowing you to focus on validation rules for {objectTypes} and manage auth at the node level. Also, a lot easier to digest for any other developers that may be involved who haven't had much interaction with Firebase. Might just have to adopt it :+1:

**Comment by DannyLingard** [comment]
Just realised I should probably also try adopt the redux pattern recommend by Dan Abramov, I'm using redux, so the structure somewhat aligns.

The below is mainly if you need a lookup table of ID's (allIds) to query (which I do a lot). I'm basing the allIds on the fact it must be faster to query and check snap.exists() or return the key on a table of Ids rather than the entire dictionary. The adminLookup is also only if you need to query values (separate node for .indexOn: 'value'). If you don't then the above is most likely all you need. 

Here's my expanded structure to suit my needs
- /adminById/{objectType}/{objectKey}
- /adminAllIds/{objectType}/{objectKey}
- /adminLookup/{objectType}/{objectKey}
- /users/{uid}
- /userOwnedById/{objectType}/{objectKey}
- /userOwnedAllIds/{objectType}/{objectKey}
- /userWriteable/{objectType}/{uid}
- /userReadableById/{objectType}/{uid}
- /userReadableAllIds/{objectType}/{uid}

As mentioned previously, I want userReadable to be a specific cut of the data I want visible to the client which is independent to userOwned, Due to this, I'll query it a lot for various checks and balances which is why I have the byId and allIds on that ref as well.

**Client Side:** For my ref listeners, I attach child listeners to the byId for each objectType within `userReadable` as well as a 'once' value event for initial data load. 

As the value event is guaranteed to fire after the child added (for existing children), I can use Redux to provide a byId and allIds for each {objectType}. I use a flag to ignore the initial child added events until the value event fires. For the value event, I use Object.keys() to create my allIds list and at the same time loops through the keys to send the snapshots through a simple object schema before populating byId. Leveraging the benefits from the pushKey being sortable, I created selectors on my redux connected components to define sort order on allIds.

So even though my redux store doesn't exactly match the sort order of my Firebase Ref (given well documented limitations in the value event) I've found it to be easier than the child_added for initial data saturation when combining things like redux-persist.

**Comment by deltaepsilon** [comment]
Sounds good. My only comment is that you might be able to do a limitToLast(1) on your child_added listener to avoid loading too much data on the initial load. The downside is that you'll get a single child_added event for an existing child after which you'll only receive child_added for new children.
